### PR TITLE
Update to stratigility 3 and move to PSR middlewares

### LINKED ISSUE #9: Support for laminas-stratigility v3
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

This package lacks support for `laminas/laminas-stratigility` v3.

### PR DESCRIPTION
**Pull request by Xerkus**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | yes
| New Feature   | yes
| RFC           | no
| QA            | no

### Description
This PR updates stratigility to 3 and enables full PSR middleware support while dropping interop middlewares support.

Backward incompatible changes include the way middleware could be defined in route config:
- New `Laminas\Mvc\Middleware\PipeSpec` value object is introduced to hold pipeline definition instead of array to avoid config merge uncertainty with array of middleware. PipeSpec is compatible with config caching via `__set_state()` magic method.
- Route match parameter `controller` must be defined along with `middleware` parameter and it's value must be a marker `Laminas\Mvc\Middleware\PipeSpec`. This serves several purposes:
  - MVC is controller based application and middleware serves as a controller alternative. Explicitly defined controller key will correctly set expectation as a controller replacement instead of a middleware in front of controller.
  - Unrelated `middleware` parameter defined for a route will not hijack execution from controller. This also mitigates same security concerns we solved by splitting controllers into their own controller manager.
  - Child route defining `controller` parameter will correctly override parent route middleware and execute that controller.
- Only acceptable values for `middleware` parameter are string container id pointing to PSR middleware or handler, instance of PSR middleware, instance of PSR handler or instance of PipeSpec. 
- Object implementing both PSR middleware and PSR handler interfaces is consumed as handler when used as a direct value of `middleware` parameter (including from container) and as a middleware when used as part of a pipe (when used in PipeSpec).
  Stratigility `Laminas\Stratigility\MiddlewarePipe` is an example of such object, however it's handler behaviour is to use EmptyPipelineHandler as a final handler.

Example configuration variants:
```php
use Laminas\Router\Http\Literal;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\MiddlewareInterface;
use Psr\Http\Server\RequestHandlerInterface;

return [
    'router' => [
        'routes' => [
            'container-middleware' => [
                'type' => Literal::class,
                'options' => [
                    'route' => '/container-middleware',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => 'container_id_for_middleware',
                    ],
                ],
            ],
            'container-handler' => [
                'type' => Literal::class,
                'options' => [
                    'route' => '/container-handler',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => 'container_id_for_handler',
                    ],
                ],
            ],
            'container-pipe' => [
                'type' => Literal::class,
                'options' => [
                    'route' => '/container-pipe',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => 'container_id_for_configured_stratigility_middlewarepipe',
                    ],
                ],
            ],
            'instance-middleware' => [
                'type' => Literal::class,
                'options' => [
                    'route' => '/instance-middleware',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => new class implements MiddlewareInterface {
                            public function process(
                                ServerRequestInterface $request,
                                RequestHandlerInterface $handler
                            ): ResponseInterface {
                                // Attempt to use $handler here will always result in empty pipeline exception 
                            }
                        },
                    ],
                ],
            ],
            'instance-handler' => [
                'type' => Literal::class,
                'options' => [
                    'route' => '/instance-handler',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => new class implements RequestHandlerInterface {
                            public function handle(
                                ServerRequestInterface $request
                            ): ResponseInterface {
                                //  
                            }
                        },
                    ],
                ],
            ],
            'instance-pipespec-preferred' => [
                'type' => Literal::class,
                'options' => [
                    'route' => '/instance-pipespec-preferred',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => new PipeSpec(
                            'container_id_for_middleware',
                            'container_id_for_another_middleware',
                            'container_id_for_handler',
                        ),
                    ],
                ],
            ],
            'instance-pipespec-dev-env' => [
                'type' => Literal::class,
                'options' => [
                    'route' => '/instance-pipespec-dev-env',
                    'defaults' => [
                        'controller' => PipeSpec::class,
                        'middleware' => new PipeSpec(
                            'container_id_for_middleware',
                            new class implements MiddlewareInterface {
                                public function process(
                                    ServerRequestInterface $request,
                                    RequestHandlerInterface $handler
                                ): ResponseInterface {
                                    return $handler->handle($request);
                                }
                            },
                            static function (
                                ServerRequestInterface $request,
                                RequestHandlerInterface $handler
                            ): ResponseInterface {
                                return $handler->handle($request);
                            },
                            'container_id_for_handler',
                        ),
                    ],
                ],
            ],
        ],
    ],
];
```
Please note that usage with instances of middleware or handlers is not compatible with config cache and provided for a development convenience, container managed middleware is strongly preferred.

### CODE REVIEW

**Comment by rieschl** [comment]
As this is a BC break anyways, you could add `array` as return type in [Module::getConfig](https://github.com/Xerkus/laminas-mvc-middleware/blob/feature/stratigility-3/src/Module.php#L19).

**Comment by Xerkus** [comment]
I did not add php 8 support, psalm or coding standard v2 here to keep out unrelated changes

File: src/MiddlewareListener.php [hunk 1]
```diff
@@ -46,24 +54,23 @@ public function onDispatch(MvcEvent $event)
         }
 
         $routeMatch = $event->getRouteMatch();
+        if (! $routeMatch || $routeMatch->getParam('controller') !== PipeSpec::class) {
```
Comment (rieschl) [inline src/MiddlewareListener.php]: `if (! $routeMatch ||`
PhpStan would complain here with:

> Only booleans are allowed in a negated boolean, `null|RouteMatch` given.

So I'd write `if ($routeMatch === null ||`

File: src/MiddlewareListener.php [hunk 2]
```diff
@@ -117,51 +116,85 @@ public function onDispatch(MvcEvent $event)
...
+            return $middlewarePipeSpec;
+        }
+
         $pipe = new MiddlewarePipe();
-        $pipe->setResponsePrototype($responsePrototype);
-        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
+        if ($middlewarePipeSpec instanceof MiddlewareInterface) {
+            $pipe->pipe($middlewarePipeSpec);
+            return $pipe;
+        }
```
Comment (rieschl) [inline src/MiddlewareListener.php]: Wouldn't it be cleaner if a new `PipeSpec` instance was created which will be handled further down anyways?

File: src/MiddlewareListener.php [hunk 3]
```diff
@@ -117,51 +116,85 @@ public function onDispatch(MvcEvent $event)
...
     /**
      * Marshal a middleware not callable exception event
      *
-     * @param  string $type
-     * @param  string $middlewareName
+     * @param string $type
+     * @param string $middlewareName
+     * @param MvcEvent $event
+     * @param Application $application
+     * @param null|Throwable $exception
```
Comment (rieschl) [inline src/MiddlewareListener.php]: These params are not necessary. There are type hints anyways.

File: src/PipeSpec.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+use Closure;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Webmozart\Assert\Assert;
+
+final class PipeSpec
+{
+    /**
+     * @var array
```
Comment (rieschl) [inline src/PipeSpec.php]: `@var list<string|RequestHandlerInterface|MiddlewareInterface|Closure>`

File: src/PipeSpec.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * @param string|RequestHandlerInterface|MiddlewareInterface|Closure ...$spec
+     */
+    public function __construct(...$spec)
+    {
+        $this->spec = $spec;
+    }
+
+    public function getSpec(): array
```
Comment (rieschl) [inline src/PipeSpec.php:34]: `@return list<string|RequestHandlerInterface|MiddlewareInterface|Closure>`

File: src/PipeSpec.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        return $this->spec;
+    }
+
+    /**
+     * Support serialization via var_export
+     */
+    public static function __set_state(array $state): self
```
Comment (rieschl) [inline src/PipeSpec.php:44]: `@var array{spec: list<string|RequestHandlerInterface|MiddlewareInterface|Closure>}  $state`

File: src/MiddlewareListener.php (hunk 2 above)
Comment (Xerkus) [reply src/MiddlewareListener.php]: I thought of that, yes

File: src/PipeSpec.php (hunk 6 above)
Comment (Xerkus) [reply src/PipeSpec.php:44]: I did not bother with those, I plan to introduce psalm as a separate PR and switch to a coding standard v2

Comment (rieschl) [reply src/PipeSpec.php:44]: Yep. That's one possibility. But does it hurt to add them now? 🙂

File: src/PipeSpec.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    private $spec;
+
+    /**
+     * @param string|RequestHandlerInterface|MiddlewareInterface|Closure ...$spec
+     */
+    public function __construct(...$spec)
+    {
+        $this->spec = $spec;
+    }
```
Comment (Xerkus) [inline src/PipeSpec.php:29]: I would like to assert value types here but since it will be in the config and cached config I am somewhat cautious.

Comment (rieschl) [reply src/PipeSpec.php:29]: With a [type hint](https://github.com/laminas/laminas-mvc-middleware/pull/3#discussion_r526120010), at least static analyzers will catch a wrong type :)

File: src/PipeSpec.php (hunk 4 above)
Comment (weierophinney) [reply src/PipeSpec.php]: @rieschl That would have to be `@psalm-var` to work correctly with our phpcs rules, I think. But yes, adding it now just helps set the expectations and make it more explicit.

File: src/PipeSpec.php (hunk 6 above)
Comment (weierophinney) [reply src/PipeSpec.php:44]: @Xerkus While I think adding psalm support later is a good call, I also think adding the declarations _now_ is good, as it makes it easier to review and understand what the expectations are for properties and arguments. When you do add psalm support, it's also that much less work to do during that step.

File: src/MiddlewareListener.php [hunk 8]
```diff
@@ -6,38 +6,46 @@
...
+    public function attach(EventManagerInterface $events, $priority = 1) : void
     {
         $this->listeners[] = $events->attach(MvcEvent::EVENT_DISPATCH, [$this, 'onDispatch'], 1);
     }
 
     /**
      * Listen to the "dispatch" event
      *
-     * @return null|Response
+     * @return null|Response|mixed
```
Comment (weierophinney) [inline src/MiddlewareListener.php]: I'd replace this with just "mixed"; psalm is going to tell you to do that anyways.

File: src/MiddlewareListener.php [hunk 9]
```diff
@@ -115,53 +111,80 @@ public function onDispatch(MvcEvent $event)
...
+        if (is_string($middlewarePipeSpec)) {
+            $middlewarePipeSpec = $this->middlewareFromContainer($container, $middlewarePipeSpec);
+        }
+        if ($middlewarePipeSpec instanceof RequestHandlerInterface) {
+            return $middlewarePipeSpec;
+        }
+
+        if ($middlewarePipeSpec instanceof MiddlewareInterface) {
+            $middlewarePipeSpec = new PipeSpec($middlewarePipeSpec);
+        } elseif (! $middlewarePipeSpec instanceof PipeSpec) {
```
Comment (weierophinney) [inline src/MiddlewareListener.php]:
```suggestion
        if ($middlewarePipeSpec instanceof MiddlewareInterface) {
            $middlewarePipeSpec = new PipeSpec($middlewarePipeSpec);
        }
        
         if (! $middlewarePipeSpec instanceof PipeSpec) {
```

Avoid else/elseif whenever you can, please.

File: src/MiddlewareListener.php [hunk 10]
```diff
@@ -115,53 +111,80 @@ public function onDispatch(MvcEvent $event)
...
+                continue;
+            }
+            if ($middlewareToBePiped instanceof RequestHandlerInterface) {
+                $middlewareToBePiped = new RequestHandlerMiddleware($middlewareToBePiped);
+                $pipe->pipe($middlewareToBePiped);
+                break; // request handler will always stop pipe processing. Rest of the pipe can be ignored
+            }
+
+            throw InvalidMiddlewareException::fromMiddlewareName($middlewareName);
         }
```
Comment (weierophinney) [inline src/MiddlewareListener.php]: Extract the contents of this loop into a method. The method should return a boolean, where one of the values means execution of the foreach should `break`:

```php
foreach ($middlewarePipeSpec->getSpec() as $middlewareToBePiped) {
    if ($this->pipeMiddleware($middlewareToBePiped, $pipe, $container)) {
        break;
    }
}
```

**Review by weierophinney — CHANGES_REQUESTED** [review]
Overall, I think this approach makes sense. My primary concern will be _migration_ for existing users; I'm hoping that this can be covered in #6...

**Comment by rieschl** [comment]
> 
> 
> Overall, I think this approach makes sense. My primary concern will be _migration_ for existing users; I'm hoping that this can be covered in #6...

I tried it in our project and the migration was straight forward, because we already have helper functions for building the middleware declarations. So it was just about adding `'controller' => PipeSpec::class,` and wrap the middleware array into a `PipeSpec` class.

Our middleware and RequestHandlers already are more or less PSR-15 interfaces so I didn't change much. If that wasn't the case, the migration would be more time consuming.

I tried to outline the changes in the migration section, but feedback of the documentation is of course welcome 🙂

File: src/MiddlewareListener.php (hunk 10 above)
Comment (Xerkus) [reply src/MiddlewareListener.php]: I extracted whole pipe creation logic into a separate factory. Please take a look again.

**Comment by Xerkus** [comment]
Currently, tests for the extracted pipe factory live in the middleware listener tests. In that way the factory is tested.
PR is in its final form short for the tests that need to be extracted to cover the factory directly

File: composer.json [hunk 11]
```diff
@@ -28,19 +28,23 @@
         }
     },
     "require": {
-        "php": "^5.6 || ^7.0",
-        "http-interop/http-middleware": "^0.4.1",
-        "laminas/laminas-mvc": "^3.1",
+        "php": "^7.3",
```
Comment (weierophinney) [inline composer.json:31]: I'm assuming 8.0 support is for a later PR?

File: composer.json [hunk 12]
```diff
@@ -28,19 +28,23 @@
...
+        "laminas/laminas-mvc": "^3.2",
         "laminas/laminas-psr7bridge": "^1.2.2",
-        "laminas/laminas-servicemanager": "^3.3",
-        "laminas/laminas-stratigility": "^2.2.2p2"
+        "laminas/laminas-servicemanager": "^3.4",
+        "laminas/laminas-stratigility": "^3.3",
+        "psr/http-server-handler": "^1.0",
+        "psr/http-server-middleware": "^1.0",
```
Comment (weierophinney) [inline composer.json:37]: The  latter depends on the former, so you really only need the middleware package here.

**Review by weierophinney — COMMENTED** [review]
I left a couple comments, but nothing critical. Probably ready to merge as-is!

File: composer.json (hunk 11 above)
Comment (Xerkus) [reply composer.json:31]: Yes. psalm and php 8 as separate PR

File: composer.json (hunk 12 above)
Comment (Xerkus) [reply composer.json:37]: Handler interface was used in this package code, so I list it as direct dependency. It worked well for me in preventing issues with forward compat changes bringing new major version of implicit dependency.
I think I will put this approach on TSC agenda as a QA requirement.

Comment (rieschl) [reply composer.json:37]: https://github.com/maglnet/ComposerRequireChecker will help you with that

Comment (Xerkus) [reply composer.json:37]: And, of course, first thing I see there is Ocramius merging PRs.

Comment (rieschl) [reply composer.json:37]: it was basically his idea, if I understood correctly

**Pull request merged**